csrc/optimization/remove_empty.cpp

```diff
@@ -0,0 +1,469 @@
+    }
+    markLiveRecursive(new_tv);
+    markDeadAndMaybeRemove(old_tv);
+  }
+
+  //! Guard removeVal with calls to markDead so that we always detect removed
+  //! Vals and Exprs before derefencing them.
+  //!
+  //! Note that we only remove val if all of its definition's outputs are marked
+  //! dead.
+  void markDeadAndMaybeRemove(Val* val) {
+    markDead(val);
+    if (val->definition()) {
+      if (allOutputsDead(val->definition())) {
+        // If all other outputs are dead, it's safe to remove the definition as
+        // well as all its outputs
+        markDead(val->definition());
+        const auto outputs = val->definition()->outputs();
+        for (auto outp : outputs) {
+          fusion_->removeVal(outp);
+        }
+      }
+    } else {
+      TORCH_INTERNAL_ASSERT(
+          !val->isFusionInput(), "Refusing to remove Fusion input");
+      fusion_->removeVal(val);
+    }
+  }
+
+  //! Find whether a statement is live (i.e. not dead code)
+  //!
+  //! Inside BackwardVisitor::traverseTo, traversal_exprs_ is built, containing
+  //! all active expressions in the original graph.
+  bool isLive(Statement* stmt) {
+    return live_statements_.find(stmt) != live_statements_.end();
+  }
+
+  //! Check whether all outputs of an expression have been marked dead
+  bool allOutputsDead(Expr* expr) {
+    return std::all_of(
+        expr->outputs().begin(), expr->outputs().end(), [&](Val* outp) {
+          return !isLive(outp);
+        });
+  }
+
+  //! Check whether all uses have been marked dead
+  bool allUsesDead(Val* val) {
+    for (const auto use : val->uses()) {
+      if (isLive(use)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  //! Mark a single Statement as being alive
+  void markLive(Statement* stmt) {
+    live_statements_.insert(stmt);
+  }
+
+  //! Ensure that a Statement and its upstream Statements are alive. If it is an
+  //! Expr, ensure all its inputs are alive. If it's a Val with a definition,
+  //! recursive to the definition.
+  void markLiveRecursive(Statement* stmt) {
+    if (isLive(stmt)) {
+      return;
+    }
+    markLive(stmt);
+    if (stmt->isVal() && stmt->asVal()->definition()) {
+      markLiveRecursive(stmt);
+    } else {
+      for (const auto inp : stmt->asExpr()->inputs()) {
+        markLiveRecursive(inp);
+      }
+    }
+  }
+
+  //! Mark a Statement* as dead so that we avoid dereferencing it later
+  void markDead(Statement* stmt) {
+    live_statements_.erase(stmt);
+  }
+
+ private:
+  Fusion* fusion_;
+
+  //! Statements are marked dead by removing them from this set
+  std::unordered_set<Statement*> live_statements_;
+};
+
+//! Get a vector of the integer positions of constant zero extent axes in the
+//! input domain. This will typically be used like
+//! `emptyAxes(TensorDomain::noReductions(tv->getMaybeRFactorDomain()))`
+std::vector<size_t> emptyAxes(std::vector<IterDomain*> domain) {
+  std::vector<size_t> empty_axes;
```

Comment:
nit: Please use `int64_t` instead of `size_t`. Unsigned makes more sense here semantically, but it's really tedious to suppress sign-vs-unsigned warnings, so generally speaking `int64_t` should be always used. Old code doesn't always follow this policy, but we should incrementally update it as well.